project/apps/web/hooks/useDebounce.ts

```diff
@@ -0,0 +1,34 @@
+import { useState, useEffect, startTransition } from "react";
+
+const useDebounce = <T>(
+  value: T,
+  delay: number,
+  onDebounceStart?: () => void,
+  onDebounceEnd?: () => void,
+): T => {
+  const [debouncedValue, setDebouncedValue] = useState<T>(value);
+
+  useEffect(() => {
+    if (onDebounceStart) {
+      onDebounceStart();
+    }
+
+    const timer = setTimeout(() => {
+      startTransition(() => {
+        setDebouncedValue(value);
+
+        if (onDebounceEnd) {
+          onDebounceEnd();
+        }
+      });
+    }, delay);
```

Comment:
would something like this be easier instead then u can pass isPending to the outer component or u can pass `isPending` and `startTransition` to this component from outside

```suggestion
  const [isPending, startTransition] = useTransition();
  useEffect(() => {
    const timer = setTimeout(() => {
      startTransition(() => {
        setDebouncedValue(value);
      });
    }, delay);
```

Reply:
I chose to pass in the callback functions instead because this transition state change should update the `confirmLoading` state for `QuestionsState`.

The alternative would be to use `useEffect`, e.g.
```typescript
useEffect(() => {
  if (isPending) {
    setLoadingState(true);
  } else {
    setLoadingState(false);
  }
  
}, [isPending, setLoadingState]);
```
which i thought was a bit inelegant

Reply:
but u could just use isPending's boolean value to determine if its loading or not right? `isPending == loadingState`

Reply:
> but u could just use isPending's boolean value to determine if its loading or not right? isPending == loadingState

could you elaborate on this sry